benchmarks/accelerate_opt/main.py

```diff
@@ -0,0 +1,352 @@
+    no_decay = ["bias", "LayerNorm.weight"]
+    optimizer_grouped_parameters = [
+        {
+            "params": [
+                p
+                for n, p in model.named_parameters()
+                if not any(no_decay_str in n for no_decay_str in no_decay)
+            ],
+            "weight_decay": 0.0,
+        },
+        {
+            "params": [
+                p
+                for n, p in model.named_parameters()
+                if any(no_decay_str in n for no_decay_str in no_decay)
+            ],
+            "weight_decay": 0.0,
+        },
+    ]
+    # New Code #
+    # Creates Dummy Optimizer if `optimizer` was specified in the config file else creates Adam Optimizer
+    optimizer_cls = (
+        torch.optim.AdamW
+        if accelerator.state.deepspeed_plugin is None
+        or "optimizer" not in accelerator.state.deepspeed_plugin.deepspeed_config
+        else DummyOptim
+    )
+    optimizer = optimizer_cls(optimizer_grouped_parameters, lr=5e-5)
+
+    # Scheduler and math around the number of training steps.
+
+    # Get gradient accumulation steps from deepspeed config if available
+    if accelerator.state.deepspeed_plugin is not None:
+        gradient_accumulation_steps = (
+            accelerator.state.deepspeed_plugin.deepspeed_config[
+                "gradient_accumulation_steps"
+            ]
+        )
+
+    num_update_steps_per_epoch = math.ceil(
+        len(train_dataloader) / gradient_accumulation_steps
+    )
+    num_train_epochs = math.ceil(max_train_steps / num_update_steps_per_epoch)
+
+    if (
+        accelerator.state.deepspeed_plugin is None
+        or "scheduler" not in accelerator.state.deepspeed_plugin.deepspeed_config
+    ):
+        lr_scheduler = get_scheduler(
+            name=SchedulerType.LINEAR,
+            optimizer=optimizer,
+            num_warmup_steps=0,
+            num_training_steps=max_train_steps,
+        )
+    else:
+        lr_scheduler = DummyScheduler(
+            optimizer,
+            total_num_steps=max_train_steps,
+            warmup_num_steps=0,
+        )
+
+    # Prepare everything with our `accelerator`.
+    (
+        model,
+        optimizer,
+        train_dataloader,
+        eval_dataloader,
+        lr_scheduler,
+    ) = accelerator.prepare(
+        model, optimizer, train_dataloader, eval_dataloader, lr_scheduler
+    )
+
+    # We need to recalculate our total training steps as the size of the training dataloader may have changed.
+    num_update_steps_per_epoch = math.ceil(
+        len(train_dataloader) / gradient_accumulation_steps
+    )
+    if max_train_steps < len(train_dataloader):
+        max_train_steps = max_train_steps * gradient_accumulation_steps
+    else:
+        max_train_steps = num_train_epochs * num_update_steps_per_epoch
+
+    # Train!
+    total_batch_size = (
+        per_gpu_batch_size * accelerator.num_processes * gradient_accumulation_steps
```

Comment:
This here might be wrong. If the model is huge, and is split into 8 GPUs, then the batch size is 1, not 8.
This calculation comes from the huggingface example which was used to train bert originally, where 8 gpus = 8 model replicas = batch size of 8. Here it's different.